src/wayland/xdg_foreign/mod.rs

```diff
@@ -0,0 +1,533 @@
+//! Utilities to allow clients to export and import handles to a surface.
+//!
+//! This module provides automatic management of exporting surfaces from one client and allowing
+//! another client to import the surface using a handle. Management of the lifetimes of exports,
+//! imports and allowing imported surfaces to be set as the parent of a surface are handled by the
+//! module.
+//!
+//! # How to use it
+//!
+//! Having a functional xdg_shell global setup is required.
+//!
+//! With a valid xdg_shell global, simply initialize the xdg_foreign global using the
+//! `xdg_foreign_init` function. After that just ensure the `XdgForeignState` returned by the
+//! function is kept alive.
+//!
+//! ```no_run
+//! # extern crate wayland_server;
+//! # use smithay::wayland::shell::xdg::{xdg_shell_init, XdgRequest};
+//! # use smithay::wayland::xdg_foreign::xdg_foreign_init;
+//!
+//! # let mut display = wayland_server::Display::new();
+//! // XDG Shell init
+//! let (shell_state, _, _) = xdg_shell_init(
+//!     &mut display,
+//!     |event: XdgRequest, dispatch_data| { /* handle the shell requests here */ },
+//!     None
+//! );
+//!
+//! let (foreign_state, _, _) = xdg_foreign_init(&mut display, shell_state.clone(), None);
+//!
+//! // Good to go!
+//! ```
+
+use crate::wayland::compositor;
+use crate::wayland::shell::is_toplevel_equivalent;
+use crate::wayland::shell::legacy::WL_SHELL_SURFACE_ROLE;
+use crate::wayland::shell::xdg::ShellState;
+use rand::distributions::{Alphanumeric, DistString};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use wayland_protocols::unstable::xdg_foreign::v2::server::{
+    zxdg_exported_v2, zxdg_exporter_v2, zxdg_imported_v2, zxdg_importer_v2,
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+use super::shell;
+
+/// An invalid surface error generated when an invalid surface is exported.
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidSurface {
+    #[allow(missing_docs)]
+    #[error("The surface was dead")]
+    Dead,
+
+    #[allow(missing_docs)]
+    #[error("The surface is not a toplevel surface")]
+    NotToplevel,
+}
+
+/// Manages all exported and imported surfaces.
+#[derive(Debug)]
+pub struct XdgForeignState {
+    log: ::slog::Logger,
+    shell: Arc<Mutex<ShellState>>,
+    exports: Vec<Export>,
+}
+
+impl XdgForeignState {
+    /// Returns a reference to an export using the handle of the export.
+    pub fn is_export_alive(&self, handle: String) -> bool {
```

Comment:
Couldn't this method just take a `&str`, rather than require a `String` (and thus a cloning of the contents)?